core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -663,13 +670,49 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     }
   }
 
+  private[spark] def getOrUpdateCompactible(reader: EventLogFileReader): Option[Boolean] = {
+    try {
+      val info = listing.read(classOf[LogInfo], reader.rootPath.toString)
+      val compactible = checkEligibilityForCompaction(info, reader)
+      if (info.compactible != compactible) {
+        listing.write(info.copy(compactible = compactible))
+      }
+      compactible
+    } catch {
+      case _: NoSuchElementException => None
+    }
+  }
+
+  protected def checkEligibilityForCompaction(
+      info: LogInfo,
+      reader: EventLogFileReader): Option[Boolean] = {
+    info.compactible.orElse {
+      // This is not applied to single event log file.
+      if (reader.lastIndex.isEmpty) {
+        Some(false)
+      } else {
+        if (reader.listEventLogFiles.length > 1) {
+          // We have at least one 'complete' file to check whether the event log is eligible to
+          // compact further.
+          val rate = eventFilterRateCalculator.calculate(
```

Comment:
That's the effort of avoiding overheads on event log on cases where compaction wouldn't help, as replaying event log file is expensive.

You're right if we imagine arbitrary cases on supporting cases, but we intuitively know that compaction would help only streaming query in most cases (please correct me if I'm missing here), which should have very short execution per job. (Actually streaming query is the target of the workload for the umbrella issue.)

In streaming world, latency is already turned into sub-second (even there's claim for sub-millisecond). Spark may show seconds of latency but even a minute of run per a batch is bad even in micro-batch. Given that each event log file cannot be smaller than 10mb (we can measure the file sizes and decide how many files to analyze if we don't want to rely on this fact), the characteristic should have been shown in first file. And the characteristic will not change across event log files, which opens another optimization for "avoiding" trial for compaction if it was analyzed as "no compaction" before.

That's based on heuristic; there's also a way like measuring statistics like you suggested if we want to make the logic be smarter, but by leveraging heuristic we can simplify the logic (to become smarter) and also avoid more overheads. What do you think?

Reply:
> we intuitively know that compaction would help only streaming query in most cases

If you're going to restrict yourself to that assumption, then why bother with such a complicated approach?

You can avoid all this and just toss all the old logs files. At most, you can create a "compact file" in one pass by keeping app start, env update, and executor start / end events; everything else is uninteresting if all you're interested in is keeping the latest jobs from a streaming query.

That is much, much simpler and more efficient than what you have.

But if you're going the complicated route I think it makes sense to think a little past that one use case. e.g. a JDBC server can benefit from this. So can a long running "shell" session (whether spark-shell or something like Zeppelin). So can servers like Livy.

Reply:
I'll just consolidate filter rate calculator and compactor into one so that event filters built from first phase read can be applied to both scoring and compacting.

I just took simplest calculation when calculate "score": given the calculation would be based on heuristic (to avoid reading event logs again... that's all what I have been struggling for), improving the calculation requires experiences on dealing with bunch of cases of event logs which I honestly don't have and should take time if we require it in the scope of this PR. I guess the simple calculation would work for streaming scenario (maybe as well as long running "shell" session which runs interactive queries) so it seems OK to start from here and exclude the improvement from the scope of this PR.

And how many files/lines/bytes we should read to decide whether the app doesn't need to be analyzed further? This seems to be require another heuristic, which I'd also like to exclude from the scope of this PR. Currently I updated the logic to calculate each time where new index of event log file comes.

We may also be able to improve building event filter via incremental read per event log file which might be another point of improvement. I'll see how to deal with this, but it'd be ideal if we can also exclude it the scope of this PR.

Does the proposal work for you, or which things you consider as mandatory?

Reply:
Haven't looked at the updated code yet, but:

> And how many files/lines/bytes we should read to decide whether the app doesn't need to be analyzed further (to even skip reading first phase read)?

I think instead of thinking like that you'll need to think about what kind of state do you need to maintain between scans (so you don't have to re-scan the same log files over and over), and what kinds of things you can get rid of. For example, things like job/stage/task retention configs can make a difference here.

But agree, do that in a separate change. This one is already pretty large as is.

Reply:
Changed to update EventFilterBuilder incrementally; the result should be same as reading from the start, as EventFilterBuilder just processes the events flowing through replay listener bus. If we memorize how far we replayed, we can continue replaying from (latest replayed + 1). It will be done per log file, as the unit of compaction is file.